### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [507]:
# Dependencies and Setup
import pandas as pd
from pathlib import Path

# Files to load
school_data_to_load = Path("Resources/schools_complete.csv")
student_data_to_load = Path("Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

school_data_complete.head()

Student ID     student_name gender  year        school_name  reading_score  \
0           0     Paul Bradley      M     9  Huang High School             96   
1           1     Victor Smith      M    12  Huang High School             90   
2           2  Kevin Rodriguez      M    12  Huang High School             41   
3           3    Richard Scott      M    12  Huang High School             89   
4           4       Bonnie Ray      F     9  Huang High School             87   

   maths_score  School ID        type  size   budget  
0           94          0  Government  2917  1910635  
1           43          0  Government  2917  1910635  
2           76          0  Government  2917  1910635  
3           86          0  Government  2917  1910635  
4           69          0  Government  2917  1910635

## Local Government Area Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average maths score 

* Calculate the average reading score

* Calculate the percentage of students with a passing maths score (50 or greater)

* Calculate the percentage of students with a passing reading score (50 or greater)

* Calculate the percentage of students who passed maths **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [508]:
#Calculate the total number of schools
total_no_schools = school_data_complete['school_name'].nunique()

print(f'Total number of schools: {total_no_schools}')

Total number of schools: 15


In [509]:
#Calculate the total number of students
total_students = school_data_complete['Student ID'].nunique()

print(f'Total number of students: {total_students:,}')

Total number of students: 39,170


In [510]:
#Calculate the total budget
total_budget = school_data_complete['budget'].unique().sum()

print(f'Total budget across all schools: ${total_budget:,.2f}')

Total budget across all schools: $24,649,428.00


In [511]:
#Calculate mean maths score
mean_math_score = school_data_complete['maths_score'].mean()

print(f'The average maths score across all students = {mean_math_score:.2f}')

The average maths score across all students = 70.34


In [512]:
#Calculate mean reading score
mean_reading_score = school_data_complete['reading_score'].mean()

print(f'The average reading score across all students = {mean_reading_score:.2f}')

The average reading score across all students = 69.98


In [513]:
#Calculate the percentage of students with a passing maths score (50 or greater)

#find the number of pass marks
pass_math_df = school_data_complete['maths_score'] >= 50

pass_math_df.sum()

#percentage of students with a passing maths score (50 or greater)
pass_math_percent_df = pass_math_df.sum() / school_data_complete['Student ID'].nunique() * 100

print("Percentage of students with a passing math score: {:.2f}%".format(pass_math_percent_df))


Percentage of students with a passing math score: 86.08%


In [514]:
#Calculate the percentage of students with a passing reading score (50 or greater)

#find the number of pass marks
pass_reading_df = school_data_complete['reading_score'] >= 50

pass_reading_df.sum()

#percentage of students with a passing reading score (50 or greater)
pass_reading_percent_df = pass_reading_df.sum() / school_data_complete['Student ID'].nunique() * 100

print("Percentage of students with a passing reading score: {:.2f}%".format(pass_reading_percent_df))

Percentage of students with a passing reading score: 84.43%


In [515]:
# Calculate the percentage of students who passed math and reading (% Overall Passing)
overall_pass_amount = ((school_data_complete['maths_score'] >= 50) & (school_data_complete['reading_score'] >= 50)).sum()

pass_both_percent = overall_pass_amount / total_students * 100

print("Percentage of students who passed both Math & Reading: {:.2f}%".format(pass_both_percent))


Percentage of students who passed both Math & Reading: 72.81%


In [516]:
#Create DF to hold the above summary 
area_summary = pd.DataFrame({
    'Total Schools': total_no_schools,
    'Total Students': [f'{total_students:,}'],
    'Total Budget': [f'${total_budget:,}'],
    'Average Maths Score': [f'{mean_math_score:.2f}'],
    'Average Reading Score': [f'{mean_reading_score:.2f}'],
    '% Passing Math': [f'{pass_math_percent_df:.2f}%'],
    '% Passing Reading': [f'{pass_reading_percent_df:.2f}%'],
    '% Overall Passing': [f'{pass_both_percent:.2f}%'],
}, index=[0])

area_summary


Total Schools Total Students Total Budget Average Maths Score  \
0             15         39,170  $24,649,428               70.34   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                 69.98         86.08%            84.43%            72.81%

## School Summary

* Create an overview table that summarises key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Maths Score
  * Average Reading Score
  * % Passing Maths
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed maths **and** reading.)
  
* Create a dataframe to hold the above results

In [517]:
# Group by school name
school_name = school_data_complete.set_index('school_name').groupby(['school_name'])

In [518]:
# school types by school name
school_type = school_data.set_index('school_name')['type']

In [519]:
#  Calculate total students
total_students = school_data.set_index('school_name')['size']

In [520]:
# Total school budget
total_school_budget = school_data.set_index('school_name')['budget']

In [521]:
# per student budget
budget_per_student = (school_data.set_index('school_name')['budget']/school_data.set_index('school_name')['size'])


In [522]:
# Average Math Score
average_math_score = school_name['maths_score'].mean()

In [523]:
# Average Reading Score
average_reading_score = school_name['reading_score'].mean()

In [524]:
pass_math_percent = (school_data_complete[school_data_complete['maths_score'] >= 50] \
                     .groupby('school_name')['Student ID'].count() \
                     / total_student * 100)


In [525]:
# % Passing Reading

pass_reading_percent = school_data_complete[school_data_complete['reading_score'] >= 50] \
                    .groupby('school_name')['Student ID'].count() \
                    /total_students*100

In [526]:
# % Overall Passing (The percentage of students that passed math and reading.

overall_pass = school_data_complete[(school_data_complete['reading_score'] >= 50) & (school_data_complete['maths_score'] >= 50)].groupby('school_name')['Student ID'].count()/total_students*100


In [527]:
total_school_budget = total_school_budget.map("${:,.2f}".format)
budget_per_student = budget_per_student.map("${:,.2f}".format)


school_summary = pd.DataFrame({
    "School Type": school_type,
    "Total Students": total_students,
    "Total School Budget": total_school_budget,
    "Per Student Budget": budget_per_student,
    "Average Maths Score": average_math_score,
    "Average Reading Score": average_reading_score,
    "% Passing Maths": pass_math_percent,
    "% Passing Reading": pass_reading_percent,
    "% Overall Passing": overall_pass,
})
                         
school_summary = school_summary.rename_axis(None)

school_summary


School Type  Total Students Total School Budget  \
Bailey High School      Government            4976       $3,124,928.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Figueroa High School    Government            2949       $1,884,411.00   
Ford High School        Government            2739       $1,763,916.00   
Griffin High School    Independent            1468         $917,500.00   
Hernandez High School   Government            4635       $3,022,020.00   
Holden High School     Independent             427         $248,087.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Pena High School       Independent             962         $585,858.00   
Rodriguez High School   Government            3999       $2,547,363.00   
Shelton High School    Independent            1761       $1,056,600.00   
Thomas High School     Independent            1635       $1,043,130.00   
Wilson High School     Independent            2283       $1,319,574.00   
Wright High School     Independent            1800       $1,049,400.00   

                      Per Student Budget  Average Maths Score  \
Bailey High School               $628.00            72.352894   
Cabrera High School              $582.00            71.657158   
Figueroa High School             $639.00            68.698542   
Ford High School                 $644.00            69.091274   
Griffin High School              $625.00            71.788147   
Hernandez High School            $652.00            68.874865   
Holden High School               $581.00            72.583138   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Pena High School                 $609.00            72.088358   
Rodriguez High School            $637.00            72.047762   
Shelton High School              $600.00            72.034072   
Thomas High School               $638.00            69.581651   
Wilson High School               $578.00            69.170828   
Wright High School               $583.00            72.047222   

                       Average Reading Score  % Passing Maths  \
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  
Bailey High School             87.379421          80.084405  
Cabrera High School            89.074273          80.785791  
Figueroa High School           82.807731          67.650051  
Ford High School               82.219788          67.469880  
Griffin High School            88.487738          81.335150  
Hernandez High School          81.877023          66.364617  
Holden High School             88.524590          78.922717  
Huang High School              81.453548          66.712376  
Johnson High School            81.978576          67.191766  
Pena High School               86.590437          79.209979  
Rodriguez High School          87.396849          79.419855  
Shelton High

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [538]:
top_performing_schools = school_summary.sort_values("% Overall Passing", ascending=False).head(5)

top_performing_schools

School Type  Total Students Total School Budget  \
Griffin High School    Independent            1468         $917,500.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Bailey High School      Government            4976       $3,124,928.00   
Wright High School     Independent            1800       $1,049,400.00   
Rodriguez High School   Government            3999       $2,547,363.00   

                       Per Student Budget  Average Maths Score  \
Griffin High School                 625.0            71.788147   
Cabrera High School                 582.0            71.657158   
Bailey High School                  628.0            72.352894   
Wright High School                  583.0            72.047222   
Rodriguez High School               637.0            72.047762   

                       Average Reading Score  % Passing Maths  \
Griffin High School                71.245232        91.212534   
Cabrera High School                71.359526        90.850377   
Bailey High School                 71.008842        91.639871   
Wright High School                 70.969444        91.777778   
Rodriguez High School              70.935984        90.797699   

                       % Passing Reading  % Overall Passing  \
Griffin High School            88.487738          81.335150   
Cabrera High School            89.074273          80.785791   
Bailey High School             87.379421          80.084405   
Wright High School             86.666667          79.722222   
Rodriguez High School          87.396849          79.419855   

                      Spending Ranges (Per Student)         School Size  
Griffin High School                        $585-630  Medium (1000-2000)  
Cabrera High School                           <$585  Medium (1000-2000)  
Bailey High School                         $585-630   Large (2000-5000)  
Wright High School                            <$585  Medium (1000-2000)  
Rodriguez High School                      $630-645   Large (2000-5000)

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [537]:
bottom_performing_schools = school_summary.sort_values("% Overall Passing", ascending=True).head(5)

bottom_performing_schools

School Type  Total Students Total School Budget  \
Hernandez High School   Government            4635       $3,022,020.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Wilson High School     Independent            2283       $1,319,574.00   
Ford High School        Government            2739       $1,763,916.00   

                       Per Student Budget  Average Maths Score  \
Hernandez High School               652.0            68.874865   
Huang High School                   655.0            68.935207   
Johnson High School                 650.0            68.843100   
Wilson High School                  578.0            69.170828   
Ford High School                    644.0            69.091274   

                       Average Reading Score  % Passing Maths  \
Hernandez High School              69.186408        80.949299   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Wilson High School                 68.876916        82.785808   
Ford High School                   69.572472        82.438846   

                       % Passing Reading  % Overall Passing  \
Hernandez High School          81.877023          66.364617   
Huang High School              81.453548          66.712376   
Johnson High School            81.978576          67.191766   
Wilson High School             81.296540          67.455103   
Ford High School               82.219788          67.469880   

                      Spending Ranges (Per Student)        School Size  
Hernandez High School                      $645-680  Large (2000-5000)  
Huang High School                          $645-680  Large (2000-5000)  
Johnson High School                        $645-680  Large (2000-5000)  
Wilson High School                            <$585  Large (2000-5000)  
Ford High School                           $630-645  Large (2000-5000)

## Maths Scores by Year

* Create a table that lists the average maths score for students of each year level (9, 10, 11, 12) at each school.

  * Create a pandas series for each year. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [530]:
year_9_math = student_data.loc[student_data['year'] == 9].groupby('school_name')["maths_score"].mean()
year_10_math = student_data.loc[student_data['year'] == 10].groupby('school_name')["maths_score"].mean()
year_11_math = student_data.loc[student_data['year'] == 11].groupby('school_name')["maths_score"].mean()
year_12_math = student_data.loc[student_data['year'] == 12].groupby('school_name')["maths_score"].mean()

mean_maths_scores = pd.DataFrame({
    "Year 9": year_9_math,
    "Year 10": year_10_math,
    "Year 11": year_11_math,
    "Year 12": year_12_math,
})

mean_maths_scores = mean_maths_scores.rename_axis(None)

mean_maths_scores

Year 9    Year 10    Year 11    Year 12
Bailey High School     72.493827  71.897498  72.374900  72.675097
Cabrera High School    72.321970  72.437768  71.008299  70.604712
Figueroa High School   68.477804  68.331586  68.811001  69.325282
Ford High School       69.021609  69.387006  69.248862  68.617811
Griffin High School    72.789731  71.093596  71.692521  71.469178
Hernandez High School  68.586831  68.867156  69.154412  68.985075
Holden High School     70.543307  75.105263  71.640777  73.409639
Huang High School      69.081754  68.533246  69.431345  68.639316
Johnson High School    69.469286  67.990220  68.637730  69.287393
Pena High School       71.996364  72.396000  72.523438  71.187845
Rodriguez High School  71.940722  71.779808  72.364811  72.154626
Shelton High School    72.932075  72.506696  70.097087  72.331536
Thomas High School     69.234273  70.057007  69.657831  69.369822
Wilson High School     69.212361  69.455446  68.378965  69.787472
Wright High School     71.741176  72.179012  73.275862  70.848238

## Reading Score by Year

* Perform the same operations as above for reading scores

In [531]:
year_9_reading = student_data.loc[student_data['year'] == 9].groupby('school_name')["reading_score"].mean()
year_10_reading = student_data.loc[student_data['year'] == 10].groupby('school_name')["reading_score"].mean()
year_11_reading = student_data.loc[student_data['year'] == 11].groupby('school_name')["reading_score"].mean()
year_12_reading = student_data.loc[student_data['year'] == 12].groupby('school_name')["reading_score"].mean()

mean_reading_scores = pd.DataFrame({
    "Year 9": year_9_reading,
    "Year 10": year_10_reading,
    "Year 11": year_11_reading,
    "Year 12": year_12_reading,
})

mean_reading_scores = mean_reading_scores.rename_axis(None)

mean_reading_scores

Year 9    Year 10    Year 11    Year 12
Bailey High School     70.901920  70.848265  70.317346  72.195525
Cabrera High School    71.172348  71.328326  71.201245  71.856021
Figueroa High School   70.261682  67.677588  69.152327  69.082126
Ford High School       69.615846  68.988701  70.735964  68.849722
Griffin High School    72.026895  70.746305  72.385042  69.434932
Hernandez High School  68.477569  70.621842  68.418199  69.244136
Holden High School     71.598425  71.096491  73.310680  70.481928
Huang High School      68.670616  69.516297  68.740638  68.671795
Johnson High School    68.719286  69.295029  69.969115  67.992521
Pena High School       70.949091  72.324000  71.703125  71.513812
Rodriguez High School  70.902921  70.137500  71.424453  71.414449
Shelton High School    70.715094  69.879464  71.150485  69.070081
Thomas High School     69.672451  69.741093  70.749398  68.730769
Wilson High School     68.683043  68.412541  68.796327  69.888143
Wright High School     71.823529  71.386831  70.475862  69.821138

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Maths Score
  * Average Reading Score
  * % Passing Maths
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [532]:
#convert 'Per Student Budget' column to numeric values
school_summary['Per Student Budget'] = school_summary['Per Student Budget'].replace('[\$,]', '', regex=True).astype(float)

#create spending bins 
spending_bins = [0, 585, 630, 645, 680]
labels = ["<$585", "$585-630", "$630-645", "$645-680"]

#categorise spending based on bins
school_summary["Spending Ranges (Per Student)"] = pd.cut(school_summary["Per Student Budget"], bins=spending_bins, labels=labels)

# Group by spending ranges
school_spending_df = school_summary.groupby("Spending Ranges (Per Student)")

# Calculate mean scores per spending range
spending_math_scores = school_spending_df["Average Maths Score"].mean().round(2)
spending_reading_scores = school_spending_df["Average Reading Score"].mean().round(2)
spending_passing_math = school_spending_df["% Passing Maths"].mean().round(2)
spending_passing_reading = school_spending_df["% Passing Reading"].mean().round(2)
overall_passing_spending = school_spending_df["% Overall Passing"].mean().round(2)

# Create the spending_summary DataFrame
spending_summary = pd.DataFrame({
    "Average Maths Score": spending_math_scores,
    "Average Reading Score": spending_reading_scores,
    "% Passing Maths": spending_passing_math,
    "% Passing Reading": spending_passing_reading,
    "% Overall Passing": overall_passing_spending
})

spending_summary.index.name = "Spending Ranges (Per Student)"

spending_summary



Average Maths Score  Average Reading Score  \
Spending Ranges (Per Student)                                               
<$585                                        71.36                  70.72   
$585-630                                     72.07                  71.03   
$630-645                                     69.85                  69.84   
$645-680                                     68.88                  69.05   

                               % Passing Maths  % Passing Reading  \
Spending Ranges (Per Student)                                       
<$585                                    88.84              86.39   
$585-630                                 91.52              87.29   
$630-645                                 84.69              83.76   
$645-680                                 81.57              81.77   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                      76.72  
$585-630                                   79.88  
$630-645                                   71.00  
$645-680                                   66.76

## Scores by School Size

* Perform the same operations as above, based on school size.

In [533]:
#create per school summary df
per_school_summary = school_summary

size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

#categorise school sizes based on bins
per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], bins=size_bins, labels=labels)

# Group by size ranges
size_summary = per_school_summary.groupby("School Size")

# Calculate mean scores per spending range
size_math_scores = size_summary["Average Maths Score"].mean().round(2)
size_reading_scores = size_summary["Average Reading Score"].mean().round(2)
size_passing_math = size_summary["% Passing Maths"].mean().round(2)
size_passing_reading = size_summary["% Passing Reading"].mean().round(2)
overall_passing_size = size_summary["% Overall Passing"].mean().round(2)

# Create the size_summary DataFrame
size_summary = pd.DataFrame({
    "Average Maths Score": size_math_scores,
    "Average Reading Score": size_reading_scores,
    "% Passing Maths": size_passing_math,
    "% Passing Reading": size_passing_reading,
    "% Overall Passing": overall_passing_size
})

size_summary.index.name = "School Size"

size_summary

Average Maths Score  Average Reading Score  \
School Size                                                      
Small (<1000)                     72.34                  71.64   
Medium (1000-2000)                71.42                  70.72   
Large (2000-5000)                 69.75                  69.58   

                    % Passing Maths  % Passing Reading  % Overall Passing  
School Size                                                                
Small (<1000)                 90.81              87.56              79.07  
Medium (1000-2000)            89.85              86.71              78.04  
Large (2000-5000)             84.25              83.30              70.29

## Scores by School Type

* Perform the same operations as above, based on school type

In [534]:
# Create Type Summary DF from previous Per School Summary DF
type_summary = per_school_summary.copy()

# Group by school type
type_summary = type_summary.groupby("School Type")

# Calculate mean scores per school type
type_math_scores = type_summary["Average Maths Score"].mean().round(2)
type_reading_scores = type_summary["Average Reading Score"].mean().round(2)
type_passing_math = type_summary["% Passing Maths"].mean().round(2)
type_passing_reading = type_summary["% Passing Reading"].mean().round(2)
overall_passing_type = type_summary["% Overall Passing"].mean().round(2)

# Create the type_summary DataFrame
type_summary = pd.DataFrame({
    "Average Maths Score": type_math_scores,
    "Average Reading Score": type_reading_scores,
    "% Passing Maths": type_passing_math,
    "% Passing Reading": type_passing_reading,
    "% Overall Passing": overall_passing_type
})

type_summary.index.name = "School Type"

type_summary

Average Maths Score  Average Reading Score  % Passing Maths  \
School Type                                                                
Government                 69.83                  69.68            84.46   
Independent                71.37                  70.72            89.20   

             % Passing Reading  % Overall Passing  
School Type                                        
Government               83.59              70.70  
Independent              86.25              76.97

# Written Report

*Task: Help the school board and Mayor make strategic decisions regarding future school budgets and priorities.*

**Analysis Summary:**

As an overalll summary, across the entire Local Government Area we can see that the overall percentage of students passing Maths, Reading, and both are a sfollows: 

**Percentage of students passing Maths:** 86.08%
**Percentage of students passing Reading:** 84.43%	
**Percentage of students passing both Maths & Reading:** 72.81%  

**Top/Botton performing schools:** 
Looking at both the top 5 and bottom 5 performing schools, there doesn't seem to be a conclusive indication that the *School Type, Total School Budget,* or *Per Student Budget* have a significant impact on student results. 
The most interesting element looking at the top and bottom 5 schools relates to the *Total Students* count. While not major, it does appear to show a slight advantage to schools with less students. 

Across the top/bottom 5, there's a variance of 3,234 students which on average is around 646 more students per school across the bottom 5 schools, potentially indicating more students = lower performing school. More analysis would be required to make a conclusion here. 

**Scores By School Spending / Budget Per Student** This section doesn't look to have any clear correlation on Student Results. There's an even spread of results across spending, and budget per student.

**Looking at the Scores By School Size**, This section further points to my Top/Bottom analysis above, indicating that smaller sized schools tend to perform better. There's a clear heirachy with the smallest sized schools achieving the best results, then the middle sized schools acheiving the middle range results, and the largest schools acheiving the lowest results. 

**Looking at the Scores By School Type**, we also see that the Independent school category has higher results in every category compared to the Government school scores. Indicating that students that attend an Independent school seem to have a higher chance of recieving higher marks. 

To conclude, the best category of school based on results within the above analysis would be a small sized (ie number of students) independent school. In regards to budget strategies, it doesn't seem effective to spend more money within established schools, but instead use budget for new schools which would reduce overall volume of students across all schools and potentially improve results. 
